src/lib/SaferNames/Name.hs

```diff
@@ -4,267 +4,113 @@
--- rank-2 polymorphism, rather than using the constructors in the data
--- definition. For example:
---   magicallyCreateFreshS :: (forall (n::S). a) -> a
---   magicallyCreateFreshS x = x   -- where does `n` come from? magic!
-
--- We also have `:=>:` to represent differences between scopes with a common
--- prefix. A `Scope (n:=>:l)` means that
---   1. `Scope n` is a prefix of `Scope l`
---   2. `Scope (n:=>:l)` is the list of names by which `l` extends `n`.
-
---      x    y    z    x    w    x
---     \-----------------/\--------/
---              n           n:=>:l
---     \---------------------------/
---                    l
-
--- Note that `l` is not necessarily a *fresh* extension: in the example above, x
--- appears in `n:=>:l` even though it already appeared, twice, in `n`. We have a
--- separate type `FreshExt n l` values of which prove that `l` is a fresh
--- extension of `n`. Without a `FreshExt n l` we merely know that `n` is a
--- prefix of `l.
-
--- There are also special scopes, `VoidScope` and `UnitScope`, representing the
--- empty list and a singleton list with a particular special name. These are
--- useful in the same way that the ordinary types `Void` and `()` are useful.
-
-data S = (:=>:) S S
-       | UnitScope
-       | VoidScope
-
--- TODO: we reuse the old `Name` to make use of the GlobalName name space while
--- we're using both the old and new systems together.
--- TODO: something like this instead:
---    type Tag = T.Text
---    data RawName = RawName Tag Int deriving (Show, Eq, Ord)
-type RawName = D.Name
-
--- invariant: the raw name in `Name s` is contained in the list in the `Scope s`
-newtype Name (n::S) = UnsafeMakeName RawName
-                      deriving (Show, Eq, Ord)
-
-newtype Env (n::S) (a:: *) = UnsafeMakeEnv (LM.LazyMap RawName a)
-                             deriving (Functor, Foldable, Traversable)
-type Scope n = Env n ()
-
-emptyEnv :: Env (n:=>:n) a
-emptyEnv = UnsafeMakeEnv mempty
-
-voidScopeEnv :: Env VoidScope a
-voidScopeEnv = UnsafeMakeEnv mempty
+import SaferNames.NameCore
+import Err
 
-infixl 1 <.>
-(<.>) :: Env (n1:=>:n2) a -> Env (n2:=>:n3) a -> Env (n1:=>:n3) a
-(<.>) (UnsafeMakeEnv m1) (UnsafeMakeEnv m2) = UnsafeMakeEnv (m2 <> m1)
+-- === environments and scopes ===
+
+-- Env is like `NameMap` but it doesn't hold the full set of names, so you can
+-- make one from an ordinary function. It's effcicently extensible using
+-- an `EnvFrag`.
+data Env (v::E->E) (i::S) (o::S) where
+  Env :: (forall s. Name s i -> v s o)  -- fallback function
+      -> NameMap v (i:=>:i') o          -- names first lookup up here
+      -> Env v i' o
+
+newtype EnvFrag (v::E->E) (i::S) (i'::S) (o::S) =
+  EnvFrag (NameMap v (i:=>:i') o)
+
+-- Scopes carry
+--    1. The full set of names (suitable for generating fresh names)
+--    2. A no-shadowing guarantee (via `Distinct`)
+--    3. The payload associated with each name's static data parameter
```

Comment:
Where does the payload appear? I assumed that would be an `s`? but I don't see one in the type signature? Is this just because it's unsafe and we have to do a runtime check?

Reply:
It's `s n` for any `Name s n`. Because it's universally quantified, the type doesn't appear in the definition of `Scope`.